### PR TITLE
Fix breakpoints on ARM

### PR DESCRIPTION
**Pull request by a20012251**

### CODE REVIEW

File: Sources/Architecture/ARM/SoftwareBreakpointManager.cpp [hunk 1]
```diff
@@ -109,7 +109,22 @@ void SoftwareBreakpointManager::enumerate(
 bool SoftwareBreakpointManager::hit(Target::Thread *thread) {
   ds2::Architecture::CPUState state;
   thread->readCPUState(state);
-  return super::hit(state.pc());
+  state.setPC(state.pc() - 2);
```
Comment (sas) [inline Sources/Architecture/ARM/SoftwareBreakpointManager.cpp]: I think this whole block should happen only `#if defined(OS_WIN32)`. Linux seems to handle the PC adjustment by itself.
If it's not too annoying, can you test that behavior yourself?

File: Sources/Architecture/ARM/SoftwareBreakpointManager.cpp [hunk 2]
```diff
@@ -109,7 +109,22 @@ void SoftwareBreakpointManager::enumerate(
...
+    //
+    // Move the PC back to the instruction
+    //
+    if (thread->writeCPUState(state) != kSuccess) {
+      abort();
+    }
+
+    uint64_t ex = state.pc();
+    thread->readCPUState(state);
+    DS2ASSERT(ex == state.pc());
```
Comment (sas) [inline Sources/Architecture/ARM/SoftwareBreakpointManager.cpp]: I don't think we should call `readCPUState()` again just to check this. As @fjricci pointed out, this check/assertion is pretty useless.

File: Sources/BreakpointManager.cpp [hunk 3]
```diff
@@ -162,11 +162,6 @@ bool BreakpointManager::hit(Address const &address) {
 
   it->second.type =
       static_cast<Type>(it->second.type & ~kTypeTemporaryUntilHit);
-  if (!it->second.type) {
```
Comment (sas) [inline Sources/BreakpointManager.cpp]: Removing this is good. Can you just add a small comment before line 163 to mention that the breakpoint will be `erase`'d when we call `BreakpointManager::disable()` if its `type` field goes to 0?

**Comment by sas** [comment]
I think this PR is good overall. You are introducing a small bug for non-Windows ARM debugging, but it should be easy to fix before checking this in.

**Comment by sas** [comment]
@facebook-github-bot shipit

File: Sources/Architecture/ARM/SoftwareBreakpointManager.cpp [hunk 4]
```diff
@@ -109,7 +109,21 @@ void SoftwareBreakpointManager::enumerate(
...
   ds2::Architecture::CPUState state;
   thread->readCPUState(state);
+#if defined(OS_WIN32)
+  state.setPC(state.pc() - 2);
+  if (super::hit(state.pc())) {
+    //
+    // Move the PC back to the instruction
+    //
+    if (thread->writeCPUState(state) != kSuccess) {
+      abort();
```
Comment (fjricci) [inline Sources/Architecture/ARM/SoftwareBreakpointManager.cpp]: Is abort() better than DS2ASSERT?

**Pull request merged**